### ISSUE TITLE
Request: Provide Anaconda package for NEST

### ISSUE THREAD

**Issue by apdavison** [nest/nest-simulator#210]
Currently, installing NEST on OS X is painful, and requires Macports or Homebrew.

Many Python users on OS X now use Anaconda, since it makes installing large packages such as Numpy and Matplotlib much easier.

Having a NEST package in the Anaconda Cloud repository would make installing NEST on OS X as simple as:

```
conda install -c https://conda.anaconda.org/nest-simulator nest
```

**Comment by apeyser** [comment]
Another reason for this: #209
Apparently, Anaconda is quite popular, but will require some package wrapping because of dynamic libraries, which will be blamed on nest. Entry point for how this is done: http://conda.pydata.org/docs/build_tutorials/pkgs.html

**Comment by apeyser** [comment]
Anaconda seems to have come from the issue of building numpy correctly, with a large number of non-python dependencies, for people who would like to move a bit faster than their distributions do --- and some distributions have quite poor numpy performance because of their choice of packages to provide those dependencies (such as using netlib blas). 

Apparently by search interest, google says that python anaconda software is 1/10 as popular as the song by Nicki Minaj. The question is, how much bother is supporting an anaconda recipe?

**Comment by apeyser** [comment]
If someone is using anaconda and is willing to share their recipe, this would be a great place to put the gist.

**Comment by jougs** [comment]
@tammoippen: Could you check if this gets any simpler using your cmake port of our build system? Thanks!

**Comment by tammoippen** [comment]
Hi,

I can confirm the loading problem (#209) on a fresh Linux (Lubuntu 15.10). I first installed Anaconda (2.4.1), then all NEST dependancies. When calling `python -c "import nest"` i get the same error:

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/test/Downloads/nest2.10.build/install/lib/python2.7/site-packages/nest/__init__.py", line 52, in <module>
        from . import pynestkernel as _kernel
    ImportError: /home/test/anaconda2/bin/../lib/libreadline.so.6: undefined symbol: PC

Calling it with debug `LD_DEBUG=libs python -c "import nest"`, reveals that python loads its readline library using the `RPATH`. Another readline will not be loaded, even if you set `LD_LIBRARY_PATH`, as `RPATH` will be searched first.

    find library=libreadline.so.6 [0]; searching
       search path=/home/test/anaconda2/bin/../lib		(RPATH from file python)
        trying file=/home/test/anaconda2/bin/../lib/libreadline.so.6

@jougs The same is true for the cmake branch.

On OS X+Anaconda+NEST I do not have any such problems.
BTW: I created a Homebrew formula for NEST [here](https://github.com/tammoippen/homebrew-science/blob/nest/nest.rb). Once this is merged, it should simplify installation of NEST on OS X (also allowing for pre-build binaries, aka Bottle, eventually).

**Comment by apeyser** [comment]
So nest is being compiled without the path set to match the anaconda lib path that python uses, right? The rpath for python and pynestkernel conflict, and the rpath for pynestkernel needs to be fixed. 

If that's right, then with anaconda on linux, you'll need to set 
LDFLAGS=-Wl,-rpath=/home/test/anaconda2/bin/../lib

or you need to install nest to /home/test/anaconda2 so that the paths match.

This would also fail any other so that is common between /usr/lib and your anaconda directory.

Another solution may be to set the LD_LIBRARY_PATH=/home/test/anaconda2/lib:$LD_LIBRARY_PATH.

or again during configure:
LDFLAGS=-L/home/test/anaconda2/lib

But again, this isn't a nest thing, this is a "if you build under anaconda without a preset recipe, you really need to know what you're doing because anaconda is a really messy solution that will tend to break everything". They probably should have build anaconda as a full chroot originally, and now it should be ported to docker.

**Comment by tammoippen** [comment]
So, neither nest, pynestkernel nor the system python installation need a rpath set to work. Only the Anaconda Python has a rpath set to load the correct libpython2.7.so (otherwise the system library would be loaded...), and others (to be independent).

And the problem is not, that pynestkernel does not find a libreadline.so, but it does not find a symbol `undefined symbol: PC`. After comparing the load orders, I found, that in the case of `python -c "import nest"` one library needed by nest is not loaded: `libncursesw.so`. So, when calling it like this:

    LD_PRELOAD=/lib/x86_64-linux-gnu/libncursesw.so.5 python -c "import nest"

everything works as expected. Further, using IPython in this case also works good:

    ipython -c "import nest"

**Comment by apeyser** [comment]
@tammoippen 
Right, the system readline is linked to libncurses, but the anaconda readline is not. 

And this mess is unavoidable for systems like anaconda. Either anaconda links readline to libncurse for everyone (and just hacks this one problem...) or we add libncurses to python (which just hacks this one problem and may cause others...) or you put anything linked to anaconda libraries into the anaconda lib directory...

What a mess. And it'll continue to be a mess unless someone who uses anaconda with linux decides to share their build with the anaconda community.

**Comment by heplesser** [comment]
It looks like #323 provides a tidy solution for the readline and ncurses problems. Together with #318, this should make installing NEST more robust. Once we have a Homebrew formula for NEST it should not be too challenging to provide something similar for Linux. Managing an Anaconda package on top of this might be more work than gain, in particular since NEST is anything but a pure python package. Therefore, I wonder if we should not close this ticket.

Another issue in this context would be whether NEST installation could place PyNEST in the standard `site-packages` directory of the Python system used instead of to `$NEST_PREFIX/lib`, at least if the prefix is not explicitly given.

**Comment by apeyser** [comment]
We should probably not put PyNEST in the standard site-packages, in order to facilitate having multiple nest versions installed. Otherwise, we'd need to combine python virtualenv with NEST installation, which could lead to further issues. Has someone tried that?

**Comment by Silmathoron** [comment]
I haven't, but asking whether PyNEST should be installed in the standard ``site-packages`` or locally in ``$NEST_PREFIX/lib`` during the install seems like a good idea to simplify the life of regular users who only have one installed version of NEST...

**Comment by kingjr** [comment]
updating ipython on anaconda generated the same error on ubuntu 14.04

adding `LD_PRELOAD=/lib/x86_64-linux-gnu/libncursesw.so.5`  to .bashrc fixed the error.

**Comment by emka** [comment]
I have built NEST as Anaconda package for a lecture, available at https://anaconda.org/emka/nest-simulator (may be a little outdated now). Feel free to copy and modify [the build recipe](https://github.com/emka/conda-recipes/tree/master/nest-simulator).

**Comment by heplesser** [comment]
@emka Thank you for creating the recipe! I will have a look at it once we have NEST 2.12 out and try to update it for that.

**Comment by muffgaga** [comment]
And just for the record:

I had to take some steps to successfully build NEST 2.12.0 with Anaconda 4.4.0 on last weeks' Debian stable (i.e. Jessie ;) ):

```bash
conda install {pyNN,lazyarray,neo,gcc,readline,ncurses,gsl}
export MY_ANACONDA3_INSTALL_DIR=~/path/to/anaconda3/
export MY_NEST_INSTALL_PATH=/do/not/use/relative/paths/here
cd my-nest-sources && mkdir build && cd build
CC=${MY_ANACONDA3_INSTALL_DIR}/bin/gcc CXX=${MY_ANACONDA3_INSTALL_DIR}/bin/g++ cmake  -Dwith-python=3 -DPYTHON_INCLUDE_DIR=${MY_ANACONDA3_INSTALL_DIR}/include/python3.6m -DPYTHON_LIBRARY=${MY_ANACONDA3_INSTALL_DIR}/lib/python3.6 -DCMAKE_INSTALL_PREFIX:PATH=MY_NEST_INSTALL_PATH ../ -Dwith-gsl=${MY_ANACONDA3_INSTALL_DIR} -DREADLINE_ROOT_DIR=${MY_ANACONDA3_INSTALL_DIR} -DLTDL_ROOT_DIR=${MY_ANACONDA3_INSTALL_DIR}
```

**Comment by jougs** [comment]
@jessica-mitchell: can you please add the information from the last comment by @muffgaga to the installation documentation and comment here once this can be closed?

**Comment by mschmidt87** [comment]
I adapted the solution of @muffgaga and ran the following code:

```
conda env create -f environment.yaml
export ANACONDA3_DIR=~/anaconda3/envs/multiarea_model
conda activate $ANACONDA3_DIR
export NEST_INSTALL_DIR=/absolute/path/to/nest/nest/master_anaconda3.bld/ins
export NEST_BUILD_DIR=/absolute/path/to/nest/master_anaconda3.bld
export NEST_SRC_DIR=/absolute/path/to/nest/nest-simulator
mkdir $NEST_BUILD_DIR && mkdir $NEST_INSTALL_DIR
cd $NEST_BUILD_DIR
CC=${ANACONDA3_DIR}/bin/gcc CXX=${ANACONDA3_DIR}/bin/g++ cmake3  -Dwith-python=3 -DPYTHON_INCLUDE_DIR=${ANACONDA3_DIR}/include/python3.6m -DPYTHON_LIBRARY=${ANACONDA3_DIR}/lib/python3.6 -DCMAKE_INSTALL_PREFIX:PATH=NEST_INSTALL_DIR $NEST_SRC_DIR -Dwith-gsl=${ANACONDA3_DIR} -DREADLINE_ROOT_DIR=${ANACONDA3_DIR} -DLTDL_ROOT_DIR=${ANACONDA3_DIR}`

```

This is my `environment.yaml` file:

```
name: multiarea_model
dependencies:
 - gcc
 - gsl
 - readline
 - ncurses
 - future
 - numpy
 - matplotlib
 - pandas
 - python==3.6
 - scipy
 - pip:
   - nested_dict
   - dicthash
```

I get the following error by cmake:

    -- The CXX compiler identification is unknown
    -- The C compiler identification is unknown
    -- Check for working CXX compiler: /home/max.schmidt/anaconda3/envs/multiarea_model/bin/g++
    -- Check for working CXX compiler: /home/max.schmidt/anaconda3/envs/multiarea_model/bin/g++ -- broken
    CMake Error at /usr/share/cmake3/Modules/CMakeTestCXXCompiler.cmake:54 (message):
    The C++ compiler "/home/max.schmidt/anaconda3/envs/multiarea_model/bin/g++"
    is not able to compile a simple test program.

    It fails with the following output:

    Change Dir: /home/max.schmidt/nest/master_anaconda3.bld/CMakeFiles/CMakeTmp

    Run Build Command:"/usr/bin/gmake" "cmTC_6fc79/fast"

    /usr/bin/gmake -f CMakeFiles/cmTC_6fc79.dir/build.make
    CMakeFiles/cmTC_6fc79.dir/build

    gmake[1]: Entering directoryowever, when I execute this (on Blaustein)
    `/mnt/beegfs/home/max.schmidt/nest/master_anaconda3.bld/CMakeFiles/CMakeTmp'

    Building CXX object CMakeFiles/cmTC_6fc79.dir/testCXXCompiler.cxx.o

    /home/max.schmidt/anaconda3/envs/multiarea_model/bin/g++
    -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona
    -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2
    -pipe -o CMakeFiles/cmTC_6fc79.dir/testCXXCompiler.cxx.o -c
    /home/max.schmidt/nest/master_anaconda3.bld/CMakeFiles/CMakeTmp/testCXXCompiler.cxx

    g++: error: unrecognized command line option ‘-std=c++17’

    g++: error: unrecognized command line option ‘-fstack-protector-strong’

    g++: error: unrecognized command line option ‘-fno-plt’

    gmake[1]: *** [CMakeFiles/cmTC_6fc79.dir/testCXXCompiler.cxx.o] Error 1

    gmake[1]: Leaving directory
    `/mnt/beegfs/home/max.schmidt/nest/master_anaconda3.bld/CMakeFiles/CMakeTmp'

    gmake: *** [cmTC_6fc79/fast] Error 2

    CMake will not be able to correctly generate this project.
    Call Stack (most recent call first):
    CMakeLists.txt:25 (project)

    -- Configuring incomplete, errors occurred!

To me, this looks as if the g++ version of my conda environment is too old. The version is 4.8.5 (obtained via g++ -v) and indeed it lacks the command line option from the error message. However, the system compiler, which I had used for NEST installations so far, has version "gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC)" and it also doesn't recognize the command line option -std=c++17.

So, is this really a problem of the compiler version?

**Comment by AlexVanMeegen** [comment]
As far as I know, the conda package `gcc` is deprecated and replaced by `gcc_linux-64` and `gxx_linux-64` (if I understand the [conda doc](https://conda.io/docs/user-guide/tasks/build-packages/compiler-tools.html#) correctly). With the `environment.yaml` file
```
name: nest
dependencies:
 - gcc_linux-64
 - gxx_linux-64
 - cmake
 - cython
 - libtool
 - gsl
 - readline
 - ncurses
 - future
 - numpy
 - matplotlib
 - pandas
 - python==3.6.5
 - scipy
 - pip:
   - nested_dict
   - dicthash
```
compiling works flawlessly for me using
```
conda env create -f environment.yaml
export CONDA_DIR=~/miniconda2/envs/nest
conda activate $CONDA_DIR
export NEST_INSTALL_DIR=/absolute/path/to/nest/nest/install
export NEST_BUILD_DIR=/absolute/path/to/nest/build
export NEST_SRC_DIR=/absolute/path/to/nest/source
mkdir $NEST_BUILD_DIR && mkdir $NEST_INSTALL_DIR
cd $NEST_BUILD_DIR
cmake -DCMAKE_INSTALL_PREFIX:PATH=$NEST_INSTALL_DIR $NEST_SRC_DIR -Dwith-gsl=${CONDA_DIR} -DREADLINE_ROOT_DIR=${CONDA_DIR} -DLTDL_ROOT_DIR=${CONDA_DIR}
make
```
Besides the replacement of `gcc`, I also added `cmake`, `cython` and `libtool` to have all NEST dependencies inside the environment. Furthermore, I had to use `python==3.6.5` because `python==3.6` came with an old `readline` version that lead to linking issues.

**Comment by AlexVanMeegen** [comment]
Beyond this, we stumbled across an intriguing issue that is likely related to the conda problems discussed here. Using the attached 
[minimal_fail.tar.gz](https://github.com/nest/nest-simulator/files/2039404/minimal_fail.tar.gz), we can produce numerically inconsistent results on the same system depending on the installation procedure of NEST and the conda environment despite passing all installchecks. This is reproducible across a number of different computers and architectures (laptops, cluster).

Using the conda environment and NEST compilation & installation procedure outlined above, `python minimal_fail.py` produces the correct results. However, if we compile and install NEST using
```
conda create -n test python=3 gsl numpy scipy matplotlib nose
conda activate test
cmake -Dwith-gsl=/home/meegen/miniconda2/envs/test -DCMAKE_INSTALL_PREFIX:PATH=/install/path /path/to/NEST/src
make -j4
make install
source /install/path/bin/nest_vars.sh
make installcheck
cd /path/to/minimal_fail
python minimal_fail.py
```
the example produces wrong results. The `cmake` summary on my laptop is
```
--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : /usr/bin/cc
C compiler flags    :  -O2 -Wall -fopenmp -fdiagnostics-color=auto
C++ compiler        : /usr/bin/c++
C++ compiler flags  :  -O2 -Wall -fopenmp -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6: /home/meegen/miniconda2/envs/test/bin/python)
       Includes     : /home/meegen/miniconda2/envs/test/include/python3.6m
       Libraries    : /home/meegen/miniconda2/envs/test/lib/libpython3.6m.so

Cython bindings     : Yes (Cython 0.27.3: /home/vanmeegen/.local/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.4)
    Includes        : /home/meegen/miniconda2/envs/test/include
    Libraries       : /home/meegen/miniconda2/envs/test/lib/libgsl.so;/home/meegen/miniconda2/envs/test/lib/libgslcblas.so

Use Readline        : Yes (GNU Readline 6.3)
    Includes        : /usr/include
    Libraries       : /usr/lib/x86_64-linux-gnu/libreadline.so;/home/meegen/miniconda2/envs/test/lib/libncurses.so

Use libltdl         : Yes (LTDL )
    Includes        : /usr/include
    Libraries       : /usr/lib/x86_64-linux-gnu/libltdl.so

Use doxygen         : No
Use MPI             : No
Use MUSIC           : No
Use libneurosim     : No

--------------------------------------------------------------------------------
```
i.e. there is a weird mix of system libraries and conda libraries.

What is not at all clear is where things break and what exactly breaks. Here are the scarce findings we've got so far:

Static compilation with `-Dstatic-libraries=ON` fails for both the 'failing' and the 'working' environment with a lot of errors like
```
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libreadline.a(display.o): In function `_rl_move_cursor_relative':
(.text+0xe64): undefined reference to `tputs'
```
all caused by `libreadline`. Not using it, i.e. `-Dwith-readline=OFF`, lets NEST compile in the 'working' environment while it fails with
```
[100%] Linking CXX shared module pynestkernel.so
/usr/bin/ld: ../nest/libnest.a(neststartup.cpp.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
../nest/libnest.a: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
pynest/CMakeFiles/pynestkernel.dir/build.make:114: recipe for target 'pynest/pynestkernel.so' failed
make[2]: *** [pynest/pynestkernel.so] Error 1
CMakeFiles/Makefile2:1169: recipe for target 'pynest/CMakeFiles/pynestkernel.dir/all' failed
make[1]: *** [pynest/CMakeFiles/pynestkernel.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```
in the 'failing' environment.

A `strace python minimal_fail.py 2>&1 >/dev/null | grep libgsl` with the 'working' environment yields
```
open("/home/meegen/miniconda2/envs/multiarea_model/lib/libgsl.so.23", O_RDONLY|O_CLOEXEC) = 4
open("/home/meegen/miniconda2/envs/multiarea_model/lib/libgslcblas.so.0", O_RDONLY|O_CLOEXEC) = 4
```
while in the 'failing' environment it returns
```
open("/home/vanmeegen/nest-test/lib/python3.6/site-packages/nest/../../../../lib/libgsl.so.23", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/vanmeegen/nest-test/lib/python3.6/site-packages/nest/../../../../../../libgsl.so.23", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/meegen/miniconda2/envs/test/lib/libgsl.so.23", O_RDONLY|O_CLOEXEC) = 4
open("/home/vanmeegen/nest-test/lib/python3.6/site-packages/nest/../../../../lib/libgslcblas.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/vanmeegen/nest-test/lib/python3.6/site-packages/nest/../../../../../../libgslcblas.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/meegen/miniconda2/envs/test/lib/libgslcblas.so.0", O_RDONLY|O_CLOEXEC) = 4
```

It seems to be not only a GSL version issue since replacing version 2.4 in the 'working' environment with 2.2 does not break it.

**Comment by heplesser** [comment]
@AlexVanMeegen Any news on this one?

**Comment by steffengraber** [comment]
I have created a NEST build recipe here: https://github.com/steffengraber/conda-nest
And then there are two Anaconda packages that can be installed in this way:

`conda install -c steffengraber -c conda-forge nest=v2.16.0`

or for master
 
`conda install -c steffengraber -c conda-forge nest=master`

Feel free to copy and use, but it would be nice to receive some feedback on this.

**Comment by heplesser** [comment]
Request Bioconda to remove NEST recipe to avoid duplicity.

**Comment by apdavison** [comment]
@heplesser I don't think you mean [duplicity](https://en.oxforddictionaries.com/definition/duplicity) :-)

**Comment by jougs** [comment]
@apdavison: so you are in favor of duplicity or what?!

**Comment by apdavison** [comment]
@jougs well if you're trying to avoid it now it rather suggests you were being duplicitous in the past... :D

**Comment by jougs** [comment]
Touché!

**Comment by steffengraber** [comment]
There is now a conda package for NEST on conda-forge

The recipe is available here:
https://github.com/conda-forge/nest-simulator-feedstock/

Packages are available here: 
https://anaconda.org/conda-forge/nest-simulator

Create a conda environment and install it with:
 
`conda install -c conda-forge nest-simulator ` 

or (with MPI support):

`conda install -c conda-forge nest-simulator=*=mpi_openmpi*`
